core/src/main/scala/org/apache/spark/sql/CachedDataFrame.scala

```diff
@@ -401,13 +404,32 @@ class CachedDataFrame(snappySession: SnappySession, queryExecution: QueryExecuti
             // no processing required
             executeCollect().iterator.asInstanceOf[Iterator[R]]
           } else {
-            val rdd = getExecRDD
-            val numPartitions = rdd.getNumPartitions
-            val results = new Array[R](numPartitions)
-            sc.runJob(rdd, processPartition, 0 until numPartitions,
-              (index: Int, r: (U, Int)) =>
-                results(index) = resultHandler(index, r._1))
-            results.iterator
+            try {
+              val execRDD = getExecRDD
+              runAsJob(execRDD, processPartition, resultHandler, sc)
+            } catch {
+              case t: Throwable
+                if CachedDataFrame.isConnectorCatalogStaleException(t, snappySession) =>
+                snappySession.externalCatalog.invalidateAll()
+                SnappySession.clearAllCache()
+                try {
+                  val execution =
+                    snappySession.getContextObject[() => QueryExecution](SnappySession.ExecutionKey)
+                  execution match {
+                    case Some(exec) =>
+                      val execRDD = exec().executedPlan.execute()
+                      runAsJob(execRDD, processPartition, resultHandler, sc)
+                    case _ => throw t
+                  }
+                } catch {
+                  case t: Throwable
```

Comment:
Should we retry for a few times or indefinitely before giving up? The same strategy in other code paths as well. Otherwise, the user has to catch this exception always.

Reply:
Yes, changed code to retry multiple times in case of failure as discussed.